cpp/src/memory/host_buffer.cpp

```diff
@@ -0,0 +1,94 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <rapidsmpf/memory/host_buffer.hpp>
+
+namespace rapidsmpf {
+
+HostBuffer::HostBuffer(
+    std::size_t size, rmm::cuda_stream_view stream, rmm::host_async_resource_ref mr
+)
+    : stream_{stream}, mr_{mr} {
+    if (size > 0) {
+        auto* ptr = static_cast<std::byte*>(mr_.allocate(stream_, size));
+        span_ = std::span<std::byte>{ptr, size};
+    }
+}
+
+void HostBuffer::deallocate_async() noexcept {
+    if (!span_.empty()) {
+        mr_.deallocate(stream_, span_.data(), span_.size());
+    }
+    span_ = {};
+}
+
+HostBuffer::HostBuffer(HostBuffer&& other) noexcept
+    : stream_{other.stream_}, mr_{other.mr_}, span_{std::exchange(other.span_, {})} {}
+
+HostBuffer& HostBuffer::operator=(HostBuffer&& other) {
+    if (this != &other) {
+        RAPIDSMPF_EXPECTS(
+            span_.empty(),
+            "cannot move into an already initialized HostBuffer",
+            std::invalid_argument
+        );
+        stream_ = other.stream_;
+        mr_ = other.mr_;
+        span_ = std::exchange(other.span_, {});
+    }
+    return *this;
+}
+
+HostBuffer::~HostBuffer() noexcept {
+    deallocate_async();
```

Comment:
I think we need a `stream_.synchronize()` here.

Reply:
`HostBuffer` is stream ordered and assumes that it is backed by a stream-ordered memory resource. This is why `HostMemoryResource::deallocate()` does, in fact, synchronize the stream.